src/main/java/org/opensearch/knn/index/KNNIndexCache.java

```diff
@@ -312,38 +310,122 @@ public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throw
+                        }
+                    }
+            );
+        } catch (Exception ex) {
+            throw new RuntimeException("Unable to query the index: " + ex);
+        } finally {
+            readLock.unlock();
+        }
     }
 
     /**
      * KNNIndexCacheEntry is the value type for entries in the cache held by {@link KNNIndexCache}.
-     * It holds a reference to both the KNNIndex and the WatcherHandle so that each can be cleaned up
-     * upon expiration of the cache.
+     * It contains all information needed to interact with jni indices and the WatcherHandle so that each can be
+     * cleaned up upon expiration of the cache.
      */
     private static class KNNIndexCacheEntry {
-        private final KNNIndex knnIndex;
+        private final long indexPointer;
+        private final long size;
         private final String indexPathUrl;
         private final String esIndexName;
+        private final String engineName;
         private final WatcherHandle<FileWatcher> fileWatcherHandle;
+        private final ReadWriteLock readWriteLock;
+        private volatile boolean isClosed;
 
-        private KNNIndexCacheEntry(final KNNIndex knnIndex, final String indexPathUrl, final String esIndexName,
-                                   final WatcherHandle<FileWatcher> fileWatcherHandle) {
-            this.knnIndex = knnIndex;
+        private KNNIndexCacheEntry(final long indexPointer, final String indexPathUrl, final String esIndexName,
+                                   final WatcherHandle<FileWatcher> fileWatcherHandle, String engineName) {
+            this.indexPointer = indexPointer;
             this.indexPathUrl = indexPathUrl;
             this.esIndexName = esIndexName;
             this.fileWatcherHandle = fileWatcherHandle;
+            this.size = IndexUtil.getFileSizeInKB(indexPathUrl);
+            this.engineName = engineName;
+            this.readWriteLock = new ReentrantReadWriteLock();
+            this.isClosed = false;
+        }
+
+        private void close() {
+            Lock writeLock = readWriteLock.writeLock();
+            writeLock.lock();
+            // Autocloseable documentation recommends making close idempotent. We don't expect to doubly close
+            // but this will help prevent a crash in that situation.
+            if (this.isClosed) {
+                return;
+            }
+            try {
+                JNIService.free(this.indexPointer, engineName);
+            } finally {
+                this.isClosed = true;
+                writeLock.unlock();
+            }
+        }
+
+        private long getIndexSize() {
+            return size;
+        }
+
+        private ReadWriteLock getReadWriteLock() {
+            return readWriteLock;
         }
 
-        private KNNIndex getKnnIndex() {
-            return knnIndex;
+        private boolean isClosed() {
```

Comment:
Do we need to have private return method?

Reply:
This call is used exclusively in the cache. So private is fine. I guess I prefer to use getters as opposed to direct access to the members so that variables dont accidentally get changed, but I understand that this does not prevent a section of code in the cache from accessing directly a value and changing it, its more for convention. That being said, what are your thoughts?

Reply:
I understand. I prefer method than variable mostly outside the class. I guess it comes in preference category, you can keep it as it is.